knora/dsplib/utils/onto_validate.py

```diff
@@ -38,15 +37,138 @@ def validate_ontology(input_file_or_json: Union[str, Dict, os.PathLike]) -> bool
+                    targets = list(link_properties[cardname])
+                    if resname not in dependencies:
+                        dependencies[resname] = dict()
+                        dependencies[resname][cardname] = targets
+                    elif cardname not in dependencies[resname]:
+                        dependencies[resname][cardname] = targets
+                    else:
+                        dependencies[resname][cardname].extend(targets)
+
+    # iteratively purge dependencies from non-circular references
+    for i in range(30):
+        # remove targets that point to a resource that is not in dependencies,
+        # remove cardinalities that have no targets
+        for res, cards in dependencies.copy().items():
+            for card, targets in cards.copy().items():
+                dependencies[res][card] = [target for target in targets if target in dependencies]
+                if len(dependencies[res][card]) == 0:
+                    del dependencies[res][card]
+        # remove resources that have no cardinalities
+        dependencies = {res: cards for res, cards in dependencies.items() if len(cards) > 0}
+        # remove resources that are not pointed to by any target
+        all_targets: Set[str] = set()
+        for cards in dependencies.values():
+            for trgt in cards.values():
+                all_targets = all_targets | set(trgt)
+        dependencies = {res: targets for res, targets in dependencies.items() if res in all_targets}
+
+    # check the remaining dependencies (which are only the circular ones) if they have all 0-1 or 0-n
+    okay_cardinalities = ['0-1', '0-n']
+    notok_dependencies: dict[str, List[str]] = dict()
+    for res, cards in dependencies.items():
+        ontoname, resname = res.split(':')
+        for card in cards:
+            # the name of the cardinality could be with prepended onto, only with colon, or without anything
+            card_without_colon = card.split(':')[1]
+            card_with_colon = ':' + card_without_colon
+            card_variations = [card, card_with_colon, card_without_colon]
+            for card_variation in card_variations:
+                match = jsonpath_ng.ext.parse(
+                    f'$[?@.name == {ontoname}].resources[?@.name == {resname}].cardinalities[?@.propname == "{card_variation}"]'
+                ).find(ontos)
+                if len(match) > 0:
+                    break
+            card_numbers = match[0].value['cardinality']
+            if card_numbers not in okay_cardinalities:
+                if res not in notok_dependencies:
+                    notok_dependencies[res] = [card]
+                else:
+                    notok_dependencies[res].append(card)
+
+    if len(notok_dependencies) == 0:
+        return True
+    else:
+        print('ERROR: Your ontology contains properties derived from "hasLinkTo" that allow circular references '
```

Comment:
Here I would print out the name of the ontology (in case there are several ontologies).

Reply:
the onto name is already prefixed to the resource names